src/DssProxyActionsCharter.sol

```diff
@@ -0,0 +1,693 @@
+    ) public payable {
+        // Receives ETH amount, converts it to WETH and joins it into the vat
+        ethJoin_join(charter, ethJoin);
+        // Locks WETH amount into the CDP
+        frob(charter, GemJoinLike(ethJoin).ilk(), toInt256(msg.value), 0);
+    }
+
+    function lockGem(
+        address charter,
+        address gemJoin,
+        uint256 amt
+    ) public {
+        // Takes token amount from user's wallet and joins into the vat
+        gemJoin_join(charter, gemJoin, amt);
+        // Locks token amount into the CDP
+        frob(charter, GemJoinLike(gemJoin).ilk(), toInt256(convertTo18(gemJoin, amt)), 0);
+    }
+
+    function freeETH(
+        address charter,
+        address ethJoin,
+        uint256 wad
+    ) public {
+        bytes32 ilk = GemJoinLike(ethJoin).ilk();
+
+        // Unlocks WETH amount from the CDP
+        frob(charter, ilk, -toInt256(wad), 0);
+        // Exits WETH amount to proxy address as a token
+        CharterLike(charter).exit(ethJoin, address(this), wad);
+        // Converts WETH to ETH
+        GemJoinLike(ethJoin).gem().withdraw(wad);
+        // Sends ETH back to the user's wallet
+        msg.sender.transfer(wad);
+    }
+
+    function freeGem(
+        address charter,
+        address gemJoin,
+        uint256 amt
+    ) public {
+        bytes32 ilk = GemJoinLike(gemJoin).ilk();
+        uint256 wad = convertTo18(gemJoin, amt);
+
+        // Unlocks token amount from the CDP
+        frob(charter, ilk, -toInt256(wad), 0);
+        // Exits token amount to the user's wallet as a token
+        CharterLike(charter).exit(gemJoin, msg.sender, amt);
+    }
+
+    function exitETH(
+        address charter,
+        address ethJoin,
+        uint256 wad
+    ) public {
+        // Exits WETH amount to proxy address as a token
+        CharterLike(charter).exit(ethJoin, address(this), wad);
+        // Converts WETH to ETH
+        GemJoinLike(ethJoin).gem().withdraw(wad);
+        // Sends ETH back to the user's wallet
+        msg.sender.transfer(wad);
+    }
+
+    function exitGem(
+        address charter,
+        address gemJoin,
+        uint256 amt
+    ) public {
+        // Exits token amount to the user's wallet as a token
+        CharterLike(charter).exit(gemJoin, msg.sender, amt);
+    }
+
+    function draw(
+        address charter,
+        bytes32 ilk,
+        address jug,
+        address daiJoin,
+        uint256 wad
+    ) public {
+        address vat = CharterLike(charter).vat();
+
+        // Generates debt in the CDP
+        int256 dart = _getDrawDart(vat, jug, ilk, wad);
+        frob(charter, ilk, 0, dart);
+        // Allows adapter to access to proxy's DAI balance in the vat
+        if (VatLike(vat).can(address(this), address(daiJoin)) == 0) {
+            VatLike(vat).hope(daiJoin);
+        }
+        // Exits DAI to the user's wallet as a token
+        DaiJoinLike(daiJoin).exit(msg.sender, _getNet(vat, charter, ilk, dart, wad));
```

Comment:
Did you consider taking care of the origination fee in `_getDrawDart` and making the user always get `wad` amount of DAI?
That way we should avoid `_getNet` function at all and also would be IMO more similar to the standard proxy actions in the sense the user receives what they ask for.